exonum-java-binding-qa-service/src/test/java/com/exonum/binding/qaservice/transactions/QaTransactionConverterTest.java

```diff
@@ -51,76 +52,61 @@ void hasFactoriesForEachTransaction(QaTransaction tx) {
     short id = tx.id();
 
     assertThat(QaTransactionConverter.TRANSACTION_FACTORIES)
-            .as("No entry for transaction %s with id=%d", tx, id)
-            .containsKey(id);
+        .as("No entry for transaction %s with id=%d", tx, id)
+        .containsKey(id);
   }
 
   @Test
   void toTransactionTransactionOfAnotherService() {
-    BinaryMessage message = new Message.Builder()
-        .setServiceId((short) (QaService.ID + 1))
-        .setMessageType(QaTransaction.INCREMENT_COUNTER.id())
-        .setBody(ByteBuffer.allocate(0))
-        .setSignature(new byte[Message.SIGNATURE_SIZE])
-        .buildRaw();
+    RawTransaction tx = RawTransaction.newBuilder()
+        .serviceId((short) (QaService.ID + 1))
+        .transactionId(QaTransaction.INCREMENT_COUNTER.id())
+        .payload(bytes())
+        .build();
 
     Exception e = assertThrows(IllegalArgumentException.class,
-        () -> converter.toTransaction(message));
-    assertThat(e).hasMessageMatching("Wrong service id \\(\\d+\\), must be "
-        + QaService.ID);
+        () -> converter.toTransaction(tx));
+    assertThat(e).hasMessageMatching("This transaction \\(.+\\) does not belong "
+        + "to this service: wrong service id \\(\\d+\\), must be " + QaService.ID);
   }
 
   @Test
   void toTransactionUnknownTransaction() {
-    UnknownTx unknownTx = new UnknownTx();
-    BinaryMessage message = unknownTx.getMessage();
+    RawTransaction raw = UnknownTx.createRawTransaction();
 
     Exception e = assertThrows(IllegalArgumentException.class,
-        () -> converter.toTransaction(message));
+        () -> converter.toTransaction(raw));
     assertThat(e).hasMessageStartingWith("Unknown transaction");
   }
 
   @ParameterizedTest
-  @MethodSource("transactionMessages")
-  void toTransaction(Class<? extends Transaction> expectedType, Message messageTemplate) {
-    // Check the converter converts a valid binary message of each known type
-    // to the expected executable transaction type.
-    BinaryMessage message = new Message.Builder()
-        .mergeFrom(messageTemplate)
-        .buildRaw();
-
-    Transaction transaction = converter.toTransaction(message);
-
-    assertThat(transaction)
-        .isInstanceOf(expectedType);
+  @MethodSource("transactions")
+  void toTransaction(Class<? extends Transaction> expectedType, RawTransaction tx) {
+    Transaction transaction = converter.toTransaction(tx);
+
+    assertThat(transaction).isInstanceOf(expectedType);
   }
 
-  private static Collection<Arguments> transactionMessages() {
-    List<Arguments> transactionTemplates = Arrays.asList(
+  private static Collection<Arguments> transactions() {
```

Comment:
I'd consider simplifying it a little:

```java

  private static Stream<Arguments> transactions() {
    return Stream.of(
        new CreateCounterTx("name"),
        …
    )
        .map(tx -> arguments(tx.getClass(), tx.toRawTransaction()));
  }
```

Reply:
it won't work because `Transaction` doesn't have `toRawTransaction` method